### DISCUSSION TITLE
Cannot make uniform work on WASM build

### DISCUSSION THREAD

**Discussion by lobis** [gfx-rs/wgpu#3027] [Q&A] [answered]
I made https://github.com/lobis/fractal-explorer in order to learn rust / gpu, which is based on [this great tutorial](https://sotrh.github.io/learn-wgpu/). The latest commit at the time of writing is da6d8e1113ce8dc6b1dc6731562639d8d9f3f585.

My goal was to have the project compile as a native application (works!) and also as a web app using WASM (this doesn't work).

I think the problem is due to uniforms, I am not sure if I am using them correctly since it seems very user unfriendly (but I am not familiar with graphics programming so...). For example in order for my program to work I have to define additional members in the uniform struct to act as padding (maybe there is a cleaner solution?)

```
pub struct Uniform {
    pub mouse: [f32; 2],
    pub time: f32,
    _padding0: i32, // necessary to avoid error
    pub domain: [[f32; 2]; 2],
    pub c: [f32; 2],
    pub mandelbrot: i32,
    _padding1: i32, // necessary to avoid error
}
```

([file](https://github.com/lobis/fractal-explorer/blob/main/src/uniform.rs))

This solution seems to work for the desktop application. However the web version doesn't seem to work. (I build it via `wasm-pack build --target web`).

I get the following error on my browser console:

```
[.WebGL-00004C560356AA00] GL_INVALID_OPERATION: It is undefined behaviour to use a uniform buffer that is too small.
```

Honestly I am not sure how to debug this error, maybe it doesn't even have to do with uniforms and my WASM configuration is just wrong (first time I build for this target).

I would greatly appreciate any help as I would love to showcase my little project as a web app 🙏🏻

**Comment by kpreid** [comment, marked as answer]
I haven't been able to think of exactly what the problem might be here, but one thing I would suggest is reordering your struct members so that you need as little manual padding as possible. This not only saves memory but protects against implementation bugs. (`wgpu` has to work across many different graphics APIs which have different ideas about alignment requirements, and I've personally encountered one bug. It's still `wgpu`'s responsibility to deal with them, but that doesn't mean there aren't unfixed ones.)

The simplest way to do this (in the absence of `vec3`s) is to order the members from largest alignment to smallest.

```
struct Uniform {
   domain: mat2x2<f32>, // align 8 size 16, running total 16
   mouse: vec2<f32>,    // align 8 size 8, running total 24
   c: vec2<f32>,        // align 8 size 8, running total 32
   time: f32,           // align 4 size 4, running total 36 
   mandelbrot: i32,     // align 4 size 4, running total 40 
}
```

```rust
pub struct Uniform {
    pub domain: [[f32; 2]; 2],
    pub mouse: [f32; 2], 
    pub c: [f32; 2],  
    pub time: f32,
    pub mandelbrot: i32,
    pub _padding: [f32; 2],
}
```

The padding at the end of the Rust struct is necessary because WebGL (but not other targets; this is a “downlevel” restriction) requires the uniform buffer's overall size to be a multiple of 16 bytes.

**Reply by lobis** [reply]
Thanks for your reply.

I implemented your suggestion and the native / desktop version still works properly, as it should.

However now the browser gives me a more descriptive error:

```
fractal_explorer.js:340 panicked at 'wgpu error: Validation Error

Caused by:
    In Device::create_render_pipeline
      note: label = `Render Pipeline`
    Downlevel flags BUFFER_BINDINGS_NOT_16_BYTE_ALIGNED are required but not supported on the device.
This is not an invalid use of WebGPU: the underlying API or device does not support enough features to be a fully compliant implementation. A subset of the features can still be used. If you are running this program on native and not in a browser and wish to work around this issue, call Adapter::downlevel_properties or Device::downlevel_properties to get a listing of the features the current platform supports.
```

![image](https://user-images.githubusercontent.com/35803280/190450334-9a10f807-c01e-4ff1-b190-e936ffc4016d.png)

From reading the error I get its some kind of bug right? Is there anything I can change in my code to make it work on a browser @kpreid ? Thank you very much!

**Reply by kpreid** [reply]
Oh, now I remember [when I hit this problem myself](https://github.com/gfx-rs/wgpu/issues/2832). You need to declare the end-padding to 16 bytes *in the shader itself*, not just in the Rust code. That is, add a field like `_padding: vec2<f32>,` to the end of the WGSL `struct`, matching the Rust.

**Reply by lobis** [reply]
> Oh, now I remember [when I hit this problem myself](https://github.com/gfx-rs/wgpu/issues/2832). You need to declare the end-padding to 16 bytes _in the shader itself_, not just in the Rust code. That is, add a field like `_padding: vec2<f32>,` to the end of the WGSL `struct`, matching the Rust.

Sadly this doesn't fix the original error, with this fix I just get my original error back (`GL_INVALID_OPERATION: It is undefined behaviour to use a uniform buffer that is too small.`).

![image](https://user-images.githubusercontent.com/35803280/190457951-c42a118d-f05e-4cd3-8bd4-1fe7e61d32b1.png)

I tried googling the error and found some posts that could be helpful such as https://stackoverflow.com/questions/64625090/problems-with-uniform-buffers-in-webgl2 but I am not sure how / if its possible to apply the suggestion.

**Reply by kpreid** [reply]
If you're getting that error then the struct declarations must not be matching in some way. I don't know why, though. Perhaps someone more familiar with GPU layout issues will have some ideas.

**Reply by lobis** [reply]
> If you're getting that error then the struct declarations must not be matching in some way. I don't know why, though. Perhaps someone more familiar with GPU layout issues will have some ideas.

Thank you for your time. I posted a question on [stack overflow](https://stackoverflow.com/questions/73735162/problem-with-uniform-buffer-on-rusts-wgpu-only-for-web-target), hopefully somebody will solve it 🤞🏻

**Reply by grovesNL** [reply]
Looks like there's something wrong with how we handle the matrix in this uniform in the GL backend (possibly on WebGL only). I don't have time to look into it right now, but using a `vec4` instead should help you to workaround it for now, e.g.:

```diff
 struct Uniform {
-   domain: mat2x2<f32>, 
+   domain: vec4<f32>,
    mouse: vec2<f32>,
    c: vec2<f32>,        
    time: f32,          
    mandelbrot: i32
    padding: vec2<f32>,
 }

/* ... */

-    let domain_size: vec2<f32> = vec2<f32>(my_uniform.domain[0].y - my_uniform.domain[0].x, my_uniform.domain[1].y - my_uniform.domain[1].x);
-    let domain_center: vec2<f32> = vec2<f32>(my_uniform.domain[0].y + my_uniform.domain[0].x, my_uniform.domain[1].y + my_uniform.domain[1].x) / 2.0;
+    let domain_size: vec2<f32> = vec2<f32>(my_uniform.domain[1] - my_uniform.domain[0], my_uniform.domain[3] - my_uniform.domain[2]);
+    let domain_center: vec2<f32> = vec2<f32>(my_uniform.domain[1] + my_uniform.domain[0], my_uniform.domain[3] + my_uniform.domain[2]) / 2.0;
 ```

**Reply by lobis** [reply]
> 



> Looks like there's something wrong with how we handle the matrix in this uniform in the GL backend (possibly on WebGL only). I don't have time to look into it right now, but using a `vec4` instead should help you to workaround it for now, e.g.:
> 
> ```diff
>  struct Uniform {
> -   domain: mat2x2<f32>, 
> +   domain: vec4<f32>,
>     mouse: vec2<f32>,
>     c: vec2<f32>,        
>     time: f32,          
>     mandelbrot: i32
>     padding: vec2<f32>,
>  }
> 
> /* ... */
> 
> -    let domain_size: vec2<f32> = vec2<f32>(my_uniform.domain[0].y - my_uniform.domain[0].x, my_uniform.domain[1].y - my_uniform.domain[1].x);
> -    let domain_center: vec2<f32> = vec2<f32>(my_uniform.domain[0].y + my_uniform.domain[0].x, my_uniform.domain[1].y + my_uniform.domain[1].x) / 2.0;
> +    let domain_size: vec2<f32> = vec2<f32>(my_uniform.domain[1] - my_uniform.domain[0], my_uniform.domain[3] - my_uniform.domain[2]);
> +    let domain_center: vec2<f32> = vec2<f32>(my_uniform.domain[1] + my_uniform.domain[0], my_uniform.domain[3] + my_uniform.domain[2]) / 2.0;
> ```

Thank you very much! it works fine, this solved my issue.
